### ISSUE TITLE
Paper #R: "A Randomized Controlled Trial on the Effects of EmbeddedComputer Language Switching"

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#116]
This issue is for discussion related to PR #12 

Author contact: @muesbeck

**Comment by kstolee** [comment]
## Authors

* Merlin Drews (born Uesbeck), pmerlindrews@gmail.com, github: muesbeck
  (corresponding author)
* Cole Peterson, cole.scott.peterson@huskers.unl.edu github: clptrsn
* Bonita Sharif, bsharif@unl.edu, github: shbonita
* Andreas Stefik, stefika@gmail.com, github: andreas-stefik

**Comment by art-reviewer** [comment]
[Paper](https://github.com/muesbeck/fse20/blob/d4c28cadb16236d90fbd073c9775ea4f4864548b/submissions/reusable/uesbecketal/embedded-polyglot-NOT-CAMERA-READY.pdf)

**Comment by anonReviewer10000** [comment]
This artifact has all the core elements, but needs a little more polish before being assigned a reusable badge.  The most important need is more documentation.

Documented: The overall content of the artifact is provided, but there is not much information on how to actually run the experiment.  Exact steps should be given to reproduce the paper results.  Imagine this situation: I am a researcher seeking to reproduce the results.  I download the repository and run the script.  But, for some reason, I get different results than the paper reports.  Is it because of different software dependencies?  Or perhaps I ran the script without a necessary flag?  A doubt about the paper will linger.  Even if it seems a very simple process, the documentation should outline exact steps that were taken to get the results presented in the paper.

Consistent: The artifact adheres to the associated paper.

Complete: Without more documentation, it is very difficult to know whether all components are provided.

Exercisable: Likewise, more documentation is needed.  It looks to be exercisable, but it would be better if precise instructions were available.

**Comment by muesbeck** [comment]
I'm unclear regarding what the communication protocol for this process is supposed to be, but I'll add some more specific instructions.

**Comment by gousiosg** [comment]
It is like a GitHub repository: you add comments and update the contents of your artifacts with a PR anytime within the review time line

**Comment by gousiosg** [comment]
@HappyReviewer We would appreciate your review till Monday morning AoE.

**Comment by HappyReviewer** [comment]
The status of this artifact is for the "reusable" badge. The components of the artifact to evaluate are the analysis scripts + data, the usable application for the experiment, and tasks. The application used for the experiment is available as a Docker container.

### Documented

I followed the README as a way to navigate the submission.

### Consistent

The tests folder contains the participants tasks as well as the solutions (as well, I checked the samples folder). There is an R script with the survey responses. I didn't try to run the R script. There are some screenshots, which I was skeptical account but ultimately I found that they were actually useful to understanding what the participants did.

### Complete
I believe that the artifact is complete in that it follows the methodology and results described in the paper.

### Exercisable

I edited the docker-compose.yml (the apache2) to 31337:80 for host:container, and the system loaded up. Cool.

However, I'm also not sure the docker image is entirely self-contained: it looks like it tries to run some apt-get updates and such, which may make the image fail in the future. I didn't look too closely here, so I may be wrong. Perhaps the authors can just confirm for me. In the past I've had the docker container be pre-built (compiled by default), with optional instructions for building.

Finally, I had a somewhat different expectation here than what the authors provided. Ideally, I had thought there would be some set of scripts that allowed me to reproduce the plots in the paper and verify some of the statistics reported in 4.3. But if this isn't the expectation for artifacts I won't stand in the way either.

**Comment by muesbeck** [comment]
Thank you for your review! Sounds like I should make clear in the documentation that the R file is the code that allows you to verify the statistics and generate the graphs. The application in the Docker images is to allow for re-use of the code to run your own replication of the study.

**Comment by HappyReviewer** [comment]
Hi @muesbeck, yes you're right (I see the ggplot calls in the script)! I should have checked the R script more carefully. Perhaps under README if you clarify "R script" and put what we should expect the script to output to make it easier to trace it back to the paper.

**Comment by kstolee** [comment]
It seems like the artifact meets all the criteria of the ```functional``` badge, but isn't quite up to the quality of reusable, which requires:
> The artifacts associated with the paper are of a quality that significantly exceeds minimal functionality. That is, they have all the qualities of the Artifacts Evaluated – Functional level, but, in addition, they are very carefully documented and well-structured to the extent that reuse and repurposing is facilitated. In particular, norms and standards of the research community for artifacts of this type are strictly adhered to. 

In particular, @anonReviewer10000 identified some missing documentation and could not fully assess the completeness of the artifact. 

@HappyReviewer and @anonReviewer10000 what are your thoughts on ```functional``` for this artifact?

@muesbeck can you create a DOI to make the artifact available (I didn't see mention of one in the reviews)?

**Comment by HappyReviewer** [comment]
I think it's absolutely functional. I am still trying to determine whether the artifacts I've been asked to review in general are Reusable. My inclination is tending towards being more conservative about Reusable as: 1) being able to apply the artifact to other contexts (not just replication), or 2) whether the system is intended to be modified or extended in some way to support future work.

The "carefully documented" seems to be a distraction to try to evaluate, since many of the functional artifacts have good documentation as well, and there is also ambiguity at what resolution things should be documented.

**Comment by muesbeck** [comment]
I'm working on making the documentation more clear with regards to 1) what can be done with the artifact and 2) how to do it (including version numbers and a pretty much one command execution of the data analysis) and 3) by adding a file with the raw analysis results in case the execution does not work.

With regards to HappyReviewer's interpretation, I do think it can be repurposed rather well. The point of including the "experiment application" in the package was so that other researchers can reuse and remix our experiment application.  Be that to run the same experiment, a really similar experiment with slightly different tasks, or a modified experiment using the application as a starting point. This include being able to reuse the SQL database, SQL code in the database (i.e. for views), as well as the R script for analysis (either exactly or as a starting point).

Of course, this isn't a "click a button and be done replicating the experiment" kind of reusable, but that is inherently impossible as it is an experiment with human participants. I think, barring the previously mentioned improvements to documentation, we got really close to making the artifact as reusable as possible.

@kstolee Is the DOI in [index.md](https://github.com/researchart/fse20/blob/master/submissions/reusable/uesbecketal/index.md) inappropriate? It's 10.5281/zenodo.3870883 or rather [https://doi.org/10.5281/zenodo.3870883](https://doi.org/10.5281/zenodo.3870883).

I'm unclear about deadlines right now, could you let me know if there are any upcoming ones with regards to updates in relation to the reviews. I was hoping to maybe create a pull request later this week or on the weekend if I can manage.

**Comment by HappyReviewer** [comment]
Okay, given @muesbeck's response, I support giving this artifact the Reusable badge. The badges seem to be more biased towards conventional systems/tools, but the argument here is a reasonable one and makes sense for this type of research.

**Comment by kstolee** [comment]
@muesbeck the DOI is good - I just missed it! 

@anonReviewer10000 give the second review and updates from the authors, what is your recommendation regarding functional/reusable?

**Comment by anonReviewer10000** [comment]
I agree that this artifact now deserves the Reusable badge.

**Comment by kstolee** [comment]
@muesbeck you've earned three badges, congratulations! Please cite your artifact in your paper using the DOI in the references.

**Comment by muesbeck** [comment]
Thank you all! Will do! Once the documentation update is done, I'll update the DOI repository as well and then finally cite the new DOI in the paper 👍
